### PR TITLE
KQL verbatim string fixes

### PR DESCRIPTION
**Pull request by josh-hildred**
Fixes quote escaping and parameter handling

### CODE REVIEW

**Review by kashwy — APPROVED** [review]

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp [hunk 1]
```diff
@@ -448,26 +405,69 @@ String IParserKQLFunction::escapeSingleQuotes(const String & input)
...
             if (*(&ch + 1) != '\'')
                 output = output + '\\' + ch;
         }
-        else if (ch == '\"')
+        else if (ch == '\"' && is_outer_quote_double)
         {
             if (*(&ch + 1) != '\"')
                 output += ch;
         }
+        else if (ch == '\'' ) // && is_outer_double_quoted
```
Comment (kothiga) [inline src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:425]: Leftover debugging comment?

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp [hunk 2]
```diff
@@ -448,26 +405,69 @@ String IParserKQLFunction::escapeSingleQuotes(const String & input)
...
+                    throw DB::Exception(DB::ErrorCodes::SYNTAX_ERROR, "Internal Error: misformed vertbatim string tokens");
+                }
+            }
+            while ((pos->type == DB::TokenType::StringLiteral || pos->type == DB::TokenType::QuotedIdentifier) && 
+            ((*(pos->begin) == '\'' && !is_outer_quote_double) || (*(pos->begin) == '\"' && is_outer_quote_double)));
+            output += escapeVerbatimString(String(verbatimString.begin() +1, verbatimString.end() -1), is_outer_quote_double);
+        }
+        else
+        {
+            output += String(pos->begin +1, pos->end -1);
```
Comment (kothiga) [inline src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:467]: Might be worth running a quick clang-format on this function as a whole

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp [hunk 3]
```diff
@@ -288,11 +275,19 @@ IParserKQLFunction::getOptionalArgument(const String & function_name, DB::IParse
 
             scopes.pop();
         }
-
-        ++pos;
+        if (token_type == DB::TokenType::At)
+        {
+            arg += "'" + iterativelyEscapeString(pos) + "'";
+            ++pos;
+            ++pos;
```
Comment (kothiga) [inline src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:282]: `pos += 2;`?

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp [hunk 4]
```diff
@@ -448,19 +405,21 @@ String IParserKQLFunction::escapeSingleQuotes(const String & input)
...
     {
         if (ch == '\\')
             output = output + '\\' + ch;
         else if (ch == '\'')
         {
-            if (*(&ch + 1) != '\'')
+            if (*(&ch + 1) != '\'' && !is_outer_quote_double)
+                output = output + '\\' + ch;
+            else if (is_outer_quote_double)
                 output = output + '\\' + ch;
```
Comment (kothiga) [inline src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:420]: Does it make sense to combine these two conditions?

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp [hunk 5]
```diff
@@ -470,4 +429,41 @@ String IParserKQLFunction::escapeVerbatimString(const String & input)
...
+    {
+        if (pos->type == DB::TokenType::At)
+        {
+            ++pos;
+            if (pos->type != DB::TokenType::StringLiteral && pos->type != DB::TokenType::QuotedIdentifier)
+                throw DB::Exception(DB::ErrorCodes::SYNTAX_ERROR, "Verbatim string expected a string literal to follow @");
+            if (pos->end - pos->begin < 2 || (*(pos->begin) != '\'' && *(pos->begin) != '\"'))
+                throw DB::Exception(DB::ErrorCodes::SYNTAX_ERROR, "Internal Error: misformed vertbatim string tokens");
+            String verbatimString = "";
+            bool is_outer_quote_double = *(pos->begin) == '\"' ? true : false;
```
Comment (kothiga) [inline src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:446]: I think this ternary is a bit redundant.

``` cpp
const bool is_outer_quote_double = (*(pos->begin) == '\"');
```

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp [hunk 6]
```diff
@@ -448,19 +405,21 @@ String IParserKQLFunction::escapeSingleQuotes(const String & input)
     return output;
 }
 
-String IParserKQLFunction::escapeVerbatimString(const String & input)
+String IParserKQLFunction::escapeVerbatimString(const String & input, bool is_outer_quote_double)
```
Comment (kothiga) [inline src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:408]: `const bool`?

**Review by kothiga — COMMENTED** [review]
These are my last thoughts. Super minor changes.

File: src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp (hunk 3 above)
Comment (josh-hildred) [reply src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:282]: `TokenIterator` provides no overload for `+=`

**Pull request merged**